Objects/object.c

```diff
@@ -399,17 +402,17 @@ _Py_ExplicitMergeRefcount(PyObject *op, Py_ssize_t extra)
     Py_ssize_t shared = _Py_atomic_load_ssize_relaxed(&op->ob_ref_shared);
     do {
         refcnt = Py_ARITHMETIC_RIGHT_SHIFT(Py_ssize_t, shared, _Py_REF_SHARED_SHIFT);
-        if (_Py_REF_IS_MERGED(shared)) {
-            return refcnt;
-        }
-
         refcnt += (Py_ssize_t)op->ob_ref_local;
         refcnt += extra;
 
         new_shared = _Py_REF_SHARED(refcnt, _Py_REF_MERGED);
     } while (!_Py_atomic_compare_exchange_ssize(&op->ob_ref_shared,
                                                 &shared, new_shared));
 
+#ifdef Py_REF_DEBUG
+    _Py_AddRefTotal(_PyInterpreterState_GET(), extra);
+#endif
+
     _Py_atomic_store_uint32_relaxed(&op->ob_ref_local, 0);
```

Comment:
Do I understand this correctly that we're not racing here because we're either own the owning thread or the world is stopped for GC?

Reply:
The possible cases are:

1) We are the owning thread and gave up ownership (typical)
2) We are the queueing thread and the owning thread has exited (gh-114824)
3) We are the queueing thread and have stopped the world (due to OOM) (gh-114824)

The GC could have used this code path, but it uses a [different function](https://github.com/python/cpython/blob/5914a211ef5542edd1f792c2684e373a42647b04/Python/gc_free_threading.c#L157) that avoids atomics (because it's stopped the world). 

The reason case (2) is thread-safe is a bit subtle:
-  Queueing is done once, so only one thread will attempt it
- Although thread ids may be reused, the merge for (2) happens while holding a lock that would need to be acquired by any new thread that reuses the thread id.